fluent-bit-kinesis.go

```diff
@@ -0,0 +1,156 @@
+    appendNewline := output.FLBPluginConfigKey(ctx, "append_newline")
+    logrus.Infof("[kinesis %d] plugin parameter append_newline = %s", pluginID, appendNewline)
+
+    if stream == "" || region == "" {
+        return nil, fmt.Errorf("[kinesis %d] stream and region are required configuration parameters", pluginID)
+    }
+
+    if partitionKey == "log" {
+        return nil, fmt.Errorf("[kinesis %d] 'log' cannot be set as the partition key", pluginID)
+    }
+
+    if partitionKey == "" {
+        logrus.Infof("[kinesis %d] no partition key provided. A random one will be generated.", pluginID)
+    }
+
+    appendNL := false
+    if  strings.ToLower(appendNewline) == "true" {
+        appendNL = true
+    }
+    return kinesis.NewOutputPlugin(region, stream, dataKeys, partitionKey, roleARN, endpoint, appendNL, pluginID)
+}
+
+
+// The "export" comments have syntactic meaning
+// This is how the compiler knows a function should be callable from the C code
+
+//export FLBPluginRegister
+func FLBPluginRegister(ctx unsafe.Pointer) int {
+    return output.FLBPluginRegister(ctx, "kinesis", "Amazon Kinesis Data Streams Fluent Bit Plugin.")
+}
+
+//export FLBPluginInit
+func  FLBPluginInit(ctx unsafe.Pointer) int {
+    plugins.SetupLogger()
+    logrus.Debugf("[kinesis] Debug log level test successful")
+    err := addPluginInstance(ctx)
+    if err != nil {
+        logrus.Errorf("[kinesis] Failed to initialize plugin: %v\n", err)
+        return output.FLB_ERROR
+    }
+    return output.FLB_OK
+}
+
+//export FLBPluginFlushCtx
+func FLBPluginFlushCtx(ctx, data unsafe.Pointer, length C.int, tag *C.char) int {
+    var count int
+    var ret int
+    var record map[interface{}]interface{}
+
+    // Create Fluent Bit decoder
+    dec := output.NewDecoder(data, int(length))
+
+    kinesisOutput := getPluginInstance(ctx)
+    fluentTag := C.GoString(tag)
+    logrus.Debugf("[kinesis %d] Found logs with tag: %s\n", kinesisOutput.PluginID, fluentTag)
+
+    for {
+        //Extract Record
+        ret, _, record = output.GetRecord(dec)
+        if ret != 0 {
+            break
+        }
+
+        retCode := kinesisOutput.AddRecord(record)
+        if retCode != output.FLB_OK {
+            return retCode
+        }
+        count++
+    }
+    err := kinesisOutput.Flush()
+    if err != nil {
+        logrus.Errorf("[kinesis %d] %v\n", kinesisOutput.PluginID, err)
+        return output.FLB_ERROR
+    }
+    logrus.Debugf("[kinesis %d] Processed %d events with tag %s\n", kinesisOutput.PluginID, count, fluentTag)
+
+    return output.FLB_OK
+}
+
+//export FLBPluginExit
+func FLBPluginExit() int {
+    // Before final exit, call Flush() for all the instances of the Output Plugin
+    for i := range pluginInstances {
+        err := pluginInstances[i].Flush()
+        if err != nil {
+            logrus.Errorf("[kinesis %d] %v\n", pluginInstances[i].PluginID, err)
+        }
+    }
+
+    return output.FLB_OK
```

Comment:
Should probably return FLB_ERROR if any of the flushes failed, so that Fluent Bit can record the failure.

Reply:
👍 We should clearly define the behavior and be explicit in the doc.